provider/vsphere/environ_availzones.go

```diff
@@ -111,15 +116,27 @@ func (env *environ) parseAvailabilityZones(args environs.StartInstanceParams) ([
 			return nil, errors.Trace(err)
 		}
 	}
-	zoneInstances, err := AvailabilityZoneAllocations(env, group)
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-	logger.Infof("found %d zones: %v", len(zoneInstances), zoneInstances)
-
 	var zoneNames []string
-	for _, z := range zoneInstances {
-		zoneNames = append(zoneNames, z.ZoneName)
+	// Vsphere will misbehave if we call AvailabilityZoneAllocations with empty
+	// groups, in this case all zones shouhld be returned.
+	if len(group) != 0 {
+		zoneInstances, err := AvailabilityZoneAllocations(env, group)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		for _, z := range zoneInstances {
+			zoneNames = append(zoneNames, z.ZoneName)
+		}
+		logger.Infof("found %d zones: %v", len(zoneInstances), zoneInstances)
+	} else {
+		zoneInstances, err := AllAvailabilityZones(env)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		for _, z := range zoneInstances {
+			zoneNames = append(zoneNames, z.Name())
+		}
+		logger.Infof("found %d zones: %v", len(zoneInstances), zoneInstances)
```

Comment:
why print 'zoneInstances' rather than 'zoneNames' here? Given 'zoneNames' is the thing we consider important enough to return. Is there something more interesting in the instance itself?